### ISSUE TITLE
Problem when combining events + get

### ISSUE THREAD

**Issue by Filirom1** [swipely/docker-api#409]
How to reproduce

In one shell:

```
$ irb 
> require "docker"
> require "time"
> Docker::Event.since(Time.now.to_i) { |event| puts Docker::Container.get(event.id) if event.status == "start" }
```

In a second shell

```
docker run --rm -it centos bash
```

The following error happens in shell 1

```
D, [2016-06-06T22:52:50.419480 #5523] DEBUG -- : [:get, "/v1.16/events", {:since=>1465246370}, nil]




D, [2016-06-06T22:53:00.810653 #5523] DEBUG -- : [:get, "/v1.16/containers/e4ea2b2537f7d566f49cdea4cedfae5345ea577883ea8a1b5917e5b2e880cd16/json", {}, nil]
Docker::Container { :id => e4ea2b2537f7d566f49cdea4cedfae5345ea577883ea8a1b5917e5b2e880cd16, :connection => Docker::Connection { :url => unix:///, :options => {:socket=>"/var/run/docker.sock"} } }
D, [2016-06-06T22:53:01.153855 #5523] DEBUG -- : [:get, "/v1.16/containers/e4ea2b2537f7d566f49cdea4cedfae5345ea577883ea8a1b5917e5b2e880cd16/json", {}, nil]
Docker::Container { :id => e4ea2b2537f7d566f49cdea4cedfae5345ea577883ea8a1b5917e5b2e880cd16, :connection => Docker::Connection { :url => unix:///, :options => {:socket=>"/var/run/docker.sock"} } }
D, [2016-06-06T22:53:01.168621 #5523] DEBUG -- : [:get, "/v1.16/containers/e4ea2b2537f7d566f49cdea4cedfae5345ea577883ea8a1b5917e5b2e880cd16/json", {}, nil]
Docker::Container { :id => e4ea2b2537f7d566f49cdea4cedfae5345ea577883ea8a1b5917e5b2e880cd16, :connection => Docker::Connection { :url => unix:///, :options => {:socket=>"/var/run/docker.sock"} } }
D, [2016-06-06T22:53:01.187218 #5523] DEBUG -- : [:get, "/v1.16/containers/e4ea2b2537f7d566f49cdea4cedfae5345ea577883ea8a1b5917e5b2e880cd16/json", {}, nil]
Docker::Container { :id => e4ea2b2537f7d566f49cdea4cedfae5345ea577883ea8a1b5917e5b2e880cd16, :connection => Docker::Connection { :url => unix:///, :options => {:socket=>"/var/run/docker.sock"} } }
Excon::Errors::SocketError: closed stream (IOError)
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/socket.rb:158:in `read_nonblock'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/socket.rb:158:in `read_nonblock'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/socket.rb:35:in `read'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/response.rb:122:in `parse'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/middlewares/response_parser.rb:6:in `response_call'
        from /usr/local/share/gems/gems/docker-api-1.28.0/lib/excon/middlewares/hijack.rb:45:in `response_call'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/connection.rb:389:in `response'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/connection.rb:253:in `request'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/middlewares/idempotent.rb:26:in `error_call'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/middlewares/base.rb:10:in `error_call'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/middlewares/base.rb:10:in `error_call'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/middlewares/base.rb:10:in `error_call'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/connection.rb:273:in `rescue in request'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/connection.rb:221:in `request'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/middlewares/idempotent.rb:26:in `error_call'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/middlewares/base.rb:10:in `error_call'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/middlewares/base.rb:10:in `error_call'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/middlewares/base.rb:10:in `error_call'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/connection.rb:273:in `rescue in request'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/connection.rb:221:in `request'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/middlewares/idempotent.rb:26:in `error_call'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/middlewares/base.rb:10:in `error_call'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/middlewares/base.rb:10:in `error_call'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/middlewares/base.rb:10:in `error_call'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/connection.rb:273:in `rescue in request'
        from /usr/local/share/gems/gems/excon-0.49.0/lib/excon/connection.rb:221:in `request'
        from /usr/local/share/gems/gems/docker-api-1.28.0/lib/docker/connection.rb:40:in `request'
        from /usr/local/share/gems/gems/docker-api-1.28.0/lib/docker/connection.rb:65:in `block (2 levels) in <class:Connection>'
        from /usr/local/share/gems/gems/docker-api-1.28.0/lib/docker/event.rb:20:in `stream'
        from /usr/local/share/gems/gems/docker-api-1.28.0/lib/docker/event.rb:26:in `since'
        from (irb):41
        from /bin/irb:12:in `<main>'irb(main):042:0>
```

Do you have any idea of the reason ?

**Comment by mpalmer** [comment]
I know why this happens: I've just bumped into it and dug down.  It's because the `/events` request is still "in progress" (it's a chunked-encoding multipart response) while your block is being called.  So, when you ask "what's my container data?", you're sending a new HTTP request down an existing socket which thinks it's sending events -- so, unsurprisingly, it doesn't work well.

Now, you'd *think* that you could fix this problem by creating a second `Docker::Connection` object and passing that to the `.get`, so you're sending that request on a separate connection.  Except... NOPE!  Excon caches the file descriptors used for UNIX sockets, and re-uses an existing fd if one exists.  So even if you create a second `Docker::Connection`, if it has the same path it gets the same underlying fd.  Excon does not care that the fd is currently being used to service another request.

I've filed excon/excon#640 about this.  As I state in that bug report, I've verified that turning off thread-safe socket caching by creating your `Docker::Connection` objects by hand, does work:

    conn = Docker::Connection.new("unix:///var/run/docker.sock", thread_safe_sockets: false)
    Docker::Event.stream({}, conn) { |ev| Docker::Container.get(ev.ID, {}, conn) if ev.Type == "container" }

However this is only safe because `Docker::Connection` creates a new `Excon::Connection` object for every single request it makes; this isn't a general-purpose solution to the problem in Excon, it's a `docker-api`-specific fix.  Preferably, Excon will fix this, however if they decline for some reason, passing `thread_safe_sockets: false` to `Excon::Connection` in `Docker::Connection#resource` by default would be a possible workaround.

**Comment by mpalmer** [comment]
For anyone interested, this problem has been fixed in Excon as of 0.59.0, released today.  If you are hitting this problem, you should make sure you're running at least excon 0.59.0 (set a dependency on `'excon', '~> 0.59'`) and you should be good to go.  Not sure whether to suggest that `docker-api` make that official with a dependency of its own; I'd do it if it were my project, but reasonable people can disagree on whether it's a good idea.

**Comment by MichaelSp** [comment]
bump
